services/repository/branch.go

```diff
@@ -60,32 +61,34 @@ func GetBranch(repo *models.Repository, branch string) (*git.Branch, error) {
 
 // GetBranches returns branches from the repository, skipping skip initial branches and
 // returning at most limit branches, or all branches if limit is 0.
-func GetBranches(repo *models.Repository, skip, limit int) ([]*git.Branch, int, error) {
-	return git.GetBranchesByPath(repo.RepoPath(), skip, limit)
+func GetBranches(ctx context.Context, repo *models.Repository, skip, limit int) ([]*git.Branch, int, error) {
+	return git.GetBranchesByPath(ctx, repo.RepoPath(), skip, limit)
 }
 
 // checkBranchName validates branch name with existing repository branches
-func checkBranchName(repo *models.Repository, name string) error {
-	gitRepo, err := git.OpenRepository(repo.RepoPath())
+// FIXME: There really has to be faster mechanism than this. git cat-file --batch-check will check a branchname
+// and or tag name... (git.DefaultContext - as a marker to come back to this)
+func checkBranchName(ctx context.Context, repo *models.Repository, name string) error {
```

Comment:
This function is very inefficient. I've managed to improve it already but it could be a lot quicker.